File: cs/3_5_1_Observer.cs
using System;
using System.Collections.Generic;

namespace Behavioral.Observer
{
    /// <summary>Інтерфейс спостерігача</summary>
    interface ICusomObserver
    {
        /// <summary>Реакція спостерігача на зміну стану субєкта </summary>
        /// <param name="subject">Субєкт, за яким ведеться спостереження</param>
        void Update(ISubject subject);
    }

    /// <summary>Інтерфейс субєкта, за яким можна вести спостереження</summary>
    interface ISubject
    {
        /// <summary>підписати спостерігача</summary>
        void Attach(ICusomObserver observer);
        /// <summary>відписати  спостерігача</summary>
        void Detach(ICusomObserver observer);
        /// <summary>повідомити всіх підписаних спостерігачів</summary>
        void Notify();
        /// <value>Гетерт для того, щоб спостерігач міг отримати стан субєкта</value>
        int State { get; }
    }

    class Subject : ISubject
    {
        /// <summary> Сатн субєкта визначаємо цілим числом. Стан повинен бути публічно доступний для спостерігача</summary>
        private int _state = 0;

        /// <value> При зміні стану повідомляємо спостерігачів </value>
        public int State
        {
            get
            {
                return _state;
            }
            set
            {
                _state = value;
                Notify();
            }
        }
        private List<ICusomObserver> Observers = new List<ICusomObserver>();
        public void Attach(ICusomObserver observer)
        {
            Observers.Add(observer);
        }
        public void Detach(ICusomObserver observer)
        {
            Observers.Remove(observer);
        }
        public void Notify()
        {
            foreach (var observer in Observers)
            {
                observer.Update(this);
            }
        }
        /// <summary>Змінює стан обєкту на випадкове ціле число від 0 до 9</summary>
        public void GenerateRandomState()
        {
            State = new Random().Next(0, 10);
        }
    }

    /// <summary>Спостерігач виводить в консоль кожну зміну стану</summary>
    class ConsoleLogObserver : ICusomObserver
    {
        public void Update(ISubject subject)
        {
            Console.WriteLine($"Нове значення стану: {subject.State}");
        }
    }



    /// <summary>Спостерігач, який виводить повідомлення про парний стан</summary>
    class EvenObserver : ICusomObserver
    {
        public void Update(ISubject subject)
        {
            if (subject.State % 2 == 0)
            {
                Console.WriteLine("Парне значення стану");
            }
        }
    }
    /// <summary>Підраховує кількість станів, що задовільняють певній умові</summary>
    class CounterObserver : ICusomObserver
    {

        private int Count = 0;
        private Predicate<int> Condition;

        /// <summary>Передаємо умову як предикат</summary>
        public CounterObserver(Predicate<int> Condidtion)
        {
            this.Condition = Condidtion;
        }
        public void Update(ISubject subject)
        {
            if (Condition(subject.State))
            {
                Count++;
                Console.WriteLine($"Стан задовільнив умову {Count} разів");
            }
        }
    }

}

File: js/3_5_1_observer.js
/**
 * Базовий клас для субєктів, за якими ведеться спостереження
 */
class Observeable {
    constructor() {
        this._observers = [];
    }
    /**
     * підписати спостерігача
     * @param {object} observer  повинен містити метод update
     */
    attach(observer) {
        if (!observer && !observer.update)
            throw "Observer error";
        this._observers.push(observer);
    }
    /**
    * відписати  спостерігача
    * @param {object} observer 
    */
    detach(observer) {
        this._observers = this._observers.filter(o => o !== observer);
    }
    /**
    * повідомити всіх підписаних спостерігачів
    */
    notify() {
        this._observers.forEach(observer => observer.update(this));
    }
}

/**
 * клас зі станом за яким можна вести спостереження
 * @extends{Observeable}
 */
class Subject extends Observeable {
    constructor() {
        super();
        this._state = 0;
    }

    /**
     * повертає стан субєкта
     * @returns{number} 
     */
    get state() {
        return this._state;
    }

    /**
     * сетер значення повідомляє всіх підписників про зміну 
     * @param{number}
     */
    set state(value) {
        if (Number(value).isNaN)
            throw `${value} is not a number`;
        this._state = value;
        this.notify();
    }

    /**
     * Змінює стан обєкту на випадкове ціле число від 0 до 9
     */
    generateRandomState() {
        this.state = Math.floor(Math.random() * 10);
    }
}

/**
 * Спостерігач виводить в консоль кожну зміну стану
*/
class ConsoleLogObserver {
    update(subject) {
        console.log(`Нове значення стану: ${subject?.state}`);
    }
}

/**
 * Спостерігач, який виводить повідомлення про парний стан
*/
class EvenObserver {
    update(subject) {
        if (subject?.state % 2 == 0) {
            console.log("Парне значення стану");
        }
    }
}

/**
 * Підраховує кількість станів, що задовільняють певній умові
*/
class CounterObserver {
    /**
     * @param {function} condition предикат
     */

    constructor(condition) {
        if (typeof condition != "function")
            throw `${condition} is not a predicate`;
        this.condition = condition;
        this.count = 0;
    }

    update(subject) {
        if (this.condition(subject?.state)) {
            this.count++;
            console.log(`Стан задовільнив умову ${this.count} разів`);
        }
    }
}


export { Subject, ConsoleLogObserver, EvenObserver, CounterObserver }